decision_service/dsclientlib/logger/ds_async_batcher.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <string>
+#include <thread>
+#include <sstream>
+
+#include "ds_concurrent_queue.h"
+#include "ds_api_status.h"
+#include "../ds_error_callback_fn.h"
+
+namespace decision_service {
+  class error_callback_fn;
+
+	// This class takes uses a queue and a background thread to accumulate events, and send them by batch asynchronously.
+	// A batch is shipped with TSender::send(data)
+	template <typename TSender>
+	class async_batcher {
+
+	public:
+
+		int append(const std::string& evt, api_status* status = nullptr)
+		{
+			if (_queue.size() < _queue_max_size) {
+				_queue.push(evt);
+				return error_code::success;
+			}
+
+			//report errors
+		  std::ostringstream os;
+		  os << "Dropped event: " << evt;
+			api_status::try_update(status, error_code::background_queue_overflow, os.str().c_str());
+
+		  return error_code::background_queue_overflow;
+		}
+
+	private:
+		void timer()//the timer triggers a queue flush (run in background)
+		{
+			while (_thread_is_running)
+			{
+				std::this_thread::sleep_for(std::chrono::milliseconds(_batch_timeout_ms));
```

Comment:
It would be better if this functioned on an amortized basis.  In particular, instead of a mystery parameter using _batch_timeout_ms = the median of the last 100 upload times seems good.

Reply:
Makes sense.  Need to understand what the failure modes.  Feels like scope creep for this PR.  Will add a work item for ds_clientlib project.  Does that work?

---
In reply to: [187034482](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187034482) [](ancestors = 187034482)